woocommerce/woocommerce-gateway-stripe
# Fix Cash App Payments in live mode

Fixes #3365 

## Changes proposed in this Pull Request:

Cash App payments in live mode are currently failing with a generic failure (`Payment failed`). It wasn't immediately obvious why these payments were failing, however @csmcneill theorised that it could be missing payment method data in `confirmCashappPayment()` request based on [this example](https://docs.stripe.com/payments/cash-app-pay/accept-a-payment?web-or-mobile=web&payments-ui-type=direct-api#handle-redirect). 

After discussing this with Stripe, they suggested that our approach of confirming cash App payments on creation and then confirming via JS (`confirmCashappPayment()`) was leading to this error. 

This PR updates our payment intent and setup intent flows for Cash App to make sure we don't confirm on creation - this is similar to voucher PM. This change then had flow on effects like not being able to pass a `return_url` and `mandates` when not confirming immediately. 

<p align="center">
  <img width="600" src="https://github.com/user-attachments/assets/70000148-c376-438e-aacf-d1523c743e98">
  <img width="600" src="https://github.com/user-attachments/assets/a1261788-695e-4972-be7a-2d570b576c88">
</p>


So to summerise the changes:

- **Cash App payments and setup intents no longer confirm immediately on creation.** This is following the advice of Stripe reps here: p1725041202438109/1723789646.813549-slack-C7U3Y3VMY and p1724352624625259/1724202287.961919-slack-C9976E5MJ. 
   - They are still confirmed via JS via [`confirmCashappPayment()`](https://docs.stripe.com/js/payment_intents/confirm_cashapp_payment)
- Following that change, **Cash App payments and intents no longer provide a return URL or mandate information** in the create request as its no longer needed and Stripe don't allow passing it in this scenario. 

## Testing instructions

1. Enable the **Cash App** payment method.
1. Ensure that test mode is disabled.
1. Add a product to your cart.
1. Proceed to checkout.
1. Select the Cash App payment method.
1. Place the order.
    - On develop you will see a generic failure displayed on checkout. 
    - On this branch, the cash app modal should appear.

| Develop | This branch |
|--------|--------|
| <img width="865" alt="Screenshot 2024-08-22 at 10 10 47 AM" src="https://github.com/user-attachments/assets/400f687b-c590-4515-b31c-f5a59e4ad67a"> | <img width="495" alt="Screenshot 2024-08-22 at 10 14 35 AM" src="https://github.com/user-attachments/assets/56d71530-beff-4251-8f9c-9225d0081e15"> | 

---


**Post merge**
